core/jvm/src/main/scala/bobcats/HashPlatform.scala

```diff
@@ -16,19 +16,36 @@
 
 package bobcats
 
-import cats.effect.kernel.Async
+import cats.effect.kernel.{Async, Resource, Sync}
 import scodec.bits.ByteVector
 
 import java.security.MessageDigest
 
+private final class JavaSecurityDigest[F[_]](val hash: MessageDigest)(implicit F: Sync[F])
+    extends UnsealedDigest[F] {
+
+  override def update(data: ByteVector): F[Unit] = F.delay(hash.update(data.toByteBuffer))
+  override val reset = F.delay(hash.reset())
+  override def get: F[ByteVector] = F.delay(ByteVector.view(hash.digest()))
+}
+
 private[bobcats] trait HashCompanionPlatform {
-  implicit def forAsync[F[_]](implicit F: Async[F]): Hash[F] =
+
+  private[bobcats] def forSync[F[_]](implicit F: Sync[F]): Hash[F] =
```

Comment:
I would really, _really_ like to reduce the constraints on this up until `MonadThrow` on as many platforms as possible. From an implicit point of view I can introduce a `LowPriority` trait, but more annoying, `MessageDigest#getInstance` is impure so `Sync` might be as far as we can go.

It's a shame because conceptually it's a pure function. wdyt @armanbilge?

Reply:
Instead of using low priority tricks, because `Sync <: MonadThrow` you can just ask for `MonadThrow` and then pattern match to discover if it's actually a `Sync`

With that said ... 

> I would really, _really_ like to reduce the constraints on this up until `MonadThrow` on as many platforms as possible.

Why? I understand if this is for legacy code, but for new code we should strongly encourage users to prefer using constraints like `Hash[F]` `Crypto[F]` or whatever we decide, instead of deriving these capabilities on-the-spot.

Then, if users are passing around those domain-specific constraints, it doesn't matter if it needs `Async` or whatever, because it would only be materialized in `IOApp` anyway.

Reply:
> because Sync <: MonadThrow you can just ask for MonadThrow

I meant, isn't an `implicit def forSync` going to conflict with `forAsync` for `IO`?

> I understand if this is for legacy code, but for new code we should strongly encourage users to prefer using constraints like Hash[F] Crypto[F] or whatever we decide

I think, it's because the function is pure conceptually and it's only because we have platform limitations that it's not. If we could just statically link the functions, they would certainly just be a function; no capability needed. At the very least then, a user can do `Hash[Either[Throwable, ?]].digest(byteVector)` rather than requiring a runtime for it to work.

Maybe I'm overthinking it, for most consumers it likely will be wrapped in an `IO`.

Reply:
> I meant, isn't an `implicit def forSync` going to conflict with `forAsync` for `IO`?

Sorry, maybe I misunderstood what you are trying to do. My point is that having prioritized implicits for constraints in the same hierarchy doesn't make sense. Add an implicit for the weakest constraint, then check at runtime if it also satisfies the stronger constraint and use that to upgrade.

--- 

> I think, it's because the function is pure conceptually and it's only because we have platform limitations that it's not. If we could just statically link the functions, they would certainly just be a function; no capability needed. At the very least then, a user can do `Hash[Either[Throwable, ?]].digest(byteVector)` rather than requiring a runtime for it to work.

This is interesting :) The reason we need an `IO` runtime is because we are actually wrapping JVM "runtime" calls.

If you eliminate the JVM runtime stuff (e.g. implement `Hash` in pure Scala) then you could offer `Hash[Either[Throwable, ?]]` without an `IO` runtime.

Reply:
Yeah - I've been trying to keep the signatures in the same place and the API in-place. I guess with no consumers it doesn't matter!

-----

Definitely. Anyway, I think we've converged on keeping the `Sync` constraint for now, even if it is unsatisfying because of the patterns of usage.